src/rgw/rgw_data_sync.cc

```diff
@@ -1129,11 +1148,14 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
       if (sync_env->observer) {
         sync_env->observer->on_bucket_changed(bs.bucket.get_key());
       }
-      /* FIXME: what do do in case of error */
-      if (marker_tracker && !entry_marker.empty()) {
+      if (!marker_handle->marker.empty()) {
```

Comment:
Should we consider only update sync marker when sync_status==0?

Reply:
good question @cfanz, and i think that's essentially what this FIXME comment was asking

it would be technically correct to make this change as you suggest, so that when a bucket sync fails our datalog processing loop in `RGWDataSyncShardCR::incremental_sync()` would keep retrying that entry until it succeeded

however, that would prevent us from making progress on later datalog entries and increase the time it takes for changes in those later buckets to sync

we use sync leases so that multiple gateways can cooperate and share the work of sync, so EBUSY errors from bucket sync are quite common as you add more gateways to the zone. if one EBUSY error blocked all other gateways from advancing in the datalog, that would severely limit how many buckets these gateways could process in parallel

so we added the error_repo as an optimization here. just before this comment, we call `error_repo->append()` to write these failed bucket shards to the error_repo, and `RGWDataSyncShardCR::incremental_sync()` will continue to retry these error_repo entries until they succeed. that allows us to safely update the datalog marker here so we can continue making progress on later buckets in the meantime

does that make sense?

Reply:
@cfanz you're right on both points, thanks!

incremental_sync() will continue to make progress either way. i agree that change would improve the accuracy of 'sync status' - the only other drawback would be around restarts, where we'd have to re-process failed datalog entries - but that happens infrequently

and because of the datalog shard lease, i can't think of any way that other gateways would compete over leases for its bucket shards

would you like open a pr with that change and see how it works?

Reply:
sorry, i see that you already opened https://github.com/ceph/ceph/pull/22866 - and that depends on this pr to work correctly. have you tried testing the two in combination?

Reply:
Not so far. I will try on it.